### PR TITLE
No tracing of nop traceable fields

### PR DESCRIPTION
**Pull request by sagudev**
Alternative approach for https://github.com/servo/servo/pull/29918. Instead of wrapping all external types into traceable wrappers, this PR introduces `no_trace` attribute on `JSTraceable drive`, which skips specified fields of structs from being traced (needed to implement nop tracer). This moves handling of traceability from `trace.rs` to where traceability is actually used, making it IMO more clear. Basic functionality of derived is heavily inspired by `ignore_malloc_size_of`.

For example if we currently have:
```rust
#[derive(JSTraceable)]
struct S { // when trace is called on struct it calls trace on all fields
  js_managed: JSManagedType,
  non_js: NonJSManagedType,
}
// expending of unsafe_no_jsmanaged_fields macro
unsafe impl JSTraceable for NonJSManagedType {
    #[inline]
    unsafe fn trace(&self, _: *mut ::js::jsapi::JSTracer) {
        // Do nothing
    }
}
```
it would become:
```rust
#[derive(JSTraceable)]
struct S { // trace is called on js_managed field only
  js_managed: JSManagedType,
  #[no_trace]
  non_js: NonJSManagedType,
}
```

## (Un)resolved Problems

- [x] Partial traceable type. Consider this type: `HashMap<NoTraceable, Traceable>`. Here we only have to trace values of hashmap but the definition of tracer for hashmap reguire both K and V to be traceable. This problem can be resolved with wrapping `NoTraceable` into traceable wrapper. For this reason `NoTrace` was also introduced:
```rust
/// Wrapper type for nop traceble
///
/// SAFETY: Inner type must not impl JSTraceable
pub struct NoTrace<T>(pub T);

#[allow(unsafe_code)]
unsafe impl<T> JSTraceable for NoTrace<T> {
    #[inline]
    unsafe fn trace(&self, _: *mut ::js::jsapi::JSTracer) {}
}
```

## Safety considerations

- [x] `#[no_trace]` is put on field that actually implements `JSTraceable`. ~~This could probably be prevented with some lint (like we have for rooting) or [static asserts](https://docs.rs/static_assertions/latest/static_assertions/macro.assert_not_impl_any.html).~~ Solved in [3123dd8](https://github.com/servo/servo/pull/29926/commits/3123dd8806e8da47ad0836958c7fd1e208c43c72) using trait shenanigans modeled after [static asserts](https://docs.rs/static_assertions/latest/static_assertions/macro.assert_not_impl_any.html).
- [x] `JSTraceable` object is put inside `NoTrace`, this is now compile error using compiler plugin magic.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by sagudev** [comment]
Both [9cf8afe](https://github.com/servo/servo/pull/29926/commits/9cf8afe3eb51769d313298f6a5f4d3fdff9841ea) and [cb69017](https://github.com/servo/servo/pull/29926/commits/cb69017d77d65c81496f02c1abc2e3d3e351febb) presents mistakes that can (and did) happen (to me).

**Comment by jdm** [comment]
This is really interesting! I think my biggest concerns about this are the extra code that needs to be added when dealing with the NoTrace wrapper, and the fact that we need the new attribute in each field that uses an untraced type like PipelineId.

**Comment by sagudev** [comment]
> the fact that we need the new attribute in each field that uses an untraced type like PipelineId

We would still allowed creating nop tracers (we have them for std types) so we do not need to spam all fields.

Some types will end up with wrappers one way or another due to orphan rules. This way we can at least get away with minimal wrapper types.

**Comment by sagudev** [comment]
Try build: https://github.com/sagudev/servo/actions/runs/5369860271

**Comment by sagudev** [comment]
The only SAFETY problem remaining is `JSTraceable` inside `NoTrace` wrapper. I do not know if it is possible to block this as negative bounds are not available, but maybe the problem can be migrated using specialization (so it would be redundant, but safe to have `JSTraceable` in `NoTrace`):
```rust
unsafe impl<T: JSTraceable> JSTraceable for NoTrace<T> {
    #[inline]
    unsafe fn trace(&self, trc: *mut JSTracer) {
        self.0.trace(trc)
    }
}
```

**Comment by sagudev** [comment]
> the fact that we need the new attribute in each field that uses an untraced type like PipelineId

That type is just very unfortunate that is external and is used in a lot of places. Most types appears only once or twice in whole script crate, so in that case `#[no_trace]` makes more sense as it moves logic directly to consumer of JSTraceable.

**Comment by sagudev** [comment]
Graveyard of experiments:
`NotTraceable`: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d1ccb35d1606d0a1f6e37b5d71bf300e (will fail due to orphan rules as directly creating `EmptyTraceable` would)
Autotraits + negative impl = negative bounds: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=37bc24d6b093d0e6f8acc763c209db9f (fails due to https://github.com/rust-lang/rust/issues/68318#issuecomment-1385837621)

More reading stuff:
- https://github.com/rust-lang/rust/issues/42721
- https://gist.github.com/nikomatsakis/bbe6821b9e79dd3eb477#negative-reasoning-from-oibit-and-rfc-586

So compiler plugin is the only option to guarantee safety of NoTrace.

**Comment by sagudev** [comment]
I believe this is ready now.

Last try build: https://github.com/sagudev/servo/actions/runs/5579796060

Do note the errors in Mac are not coming from my PR but are also reproducible on master: 
https://github.com/sagudev/servo/actions/runs/5578404518/jobs/10192694562

also frequently there is failing wpt test (visible on same run): 
```
Stable unexpected results (2): 
  ▶ CRASH [expected OK] /fetch/local-network-access/nested-worker.window.html
```

**Comment by sagudev** [comment]
WPT export also have python errors like MAC.

**Comment by sagudev** [comment]
expected fail due to permission:

@bors-servo try

**Comment by sagudev** [comment]
expected fail due to permission:

@bors-servo try

**Comment by sagudev** [comment]
Okay, now back to main topic here.
Can I get review here. I would like to get this merged so I cab continue on #29918.

File: components/script_plugins/trace_in_no_trace.rs [hunk 1]
```diff
@@ -16,20 +19,17 @@ use rustc_span::symbol::Symbol;
 declare_lint!(
     TRACE_IN_NO_TRACE,
     Deny,
-    "Warn and report usage of traceable (jsmanaged) objects in NoTrace wrappers"
+    "Warn and report incorect usage of Traceable (jsmanaged) objects in must_not_have_traceable marked wrappers"
```
Comment (jdm) [inline components/script_plugins/trace_in_no_trace.rs:22]: s/incorect/incorrect/

**Review by jdm — APPROVED** [review]
Wow. Thank you for doing this work; I think we've ended up with the best solution that's possible given the constraints of the language!

File: components/jstraceable_derive/lib.rs [hunk 2]
```diff
@@ -7,14 +7,148 @@ extern crate syn;
...
+            impl<K, V> NoTraceOnJSTraceable<Invalid5> for std::collections::BTreeMap<K, V>
+            where
+                K: std::cmp::Eq + std::hash::Hash,
+                V: crate::dom::bindings::trace::JSTraceable,
+            {
+            }
+
+            // If there is only one specialized trait impl, type inference with
+            // `_` can be resolved and this can compile. Fails to compile if
+            // ty implements `NoTraceOnJSTraceable<InvalidX>`.
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/jstraceable_derive/lib.rs:138]: Yes, either this or another compiler plugin.

File: components/script/dom/document.rs [hunk 3]
```diff
@@ -5369,6 +5371,7 @@ impl PendingInOrderScriptVec {
 #[unrooted_must_root_lint::must_root]
 struct PendingScript {
     element: Dom<HTMLScriptElement>,
+    // TODO(sagudev): could this be all no_trace?
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/script/dom/document.rs:5389]: IIRC this type has nestedly all no trace types so it is de facto no trace, but using this shortcut would not be safe if something changes in any of dependent struct. This was just remaining part of prototyping that should be removed.

**Pull request merged**